File: src/main/java/me/itzisonn_/meazy/command/custom/CompileCommand.java
package me.itzisonn_.meazy.command.custom;

import me.itzisonn_.meazy.FileUtils;
import me.itzisonn_.meazy.MeazyMain;
import me.itzisonn_.meazy.Registries;
import me.itzisonn_.meazy.command.AbstractCommand;
import me.itzisonn_.meazy.lang.text.Text;
import me.itzisonn_.meazy.lexer.Token;
import me.itzisonn_.meazy.logging.LogLevel;
import me.itzisonn_.meazy.parser.ast.Program;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.List;

public class CompileCommand extends AbstractCommand {
    public CompileCommand() {
        super("compile", List.of("<file_to_compile>", "<output_file_path>"));
    }

    @Override
    public Text execute(String[] args) {
        File file = new File(args[0]);
        if (file.isDirectory() || !file.exists()) {
            MeazyMain.LOGGER.log(LogLevel.ERROR, Text.translatable("meazy:file.doesnt_exist", file.getAbsolutePath()));
            return null;
        }

        if (!FileUtils.getExtension(file).equals("mea")) {
            MeazyMain.LOGGER.log(LogLevel.ERROR, Text.translatable("meazy:file.unsupported_extension", FileUtils.getExtension(file)));
            return null;
        }

        MeazyMain.LOGGER.log(LogLevel.INFO, Text.translatable("meazy:commands.compile.compiling'", file.getAbsolutePath()));

        long startMillis = System.currentTimeMillis();
        List<Token> tokens = Registries.TOKENIZATION_FUNCTION.getEntry().getValue().tokenize(FileUtils.getLines(file));

        Program program = Registries.PARSE_TOKENS_FUNCTION.getEntry().getValue().parse(file, tokens);
        long endMillis = System.currentTimeMillis();


        File outputFile = new File(args[1]);
        if (file.isDirectory()) {
            MeazyMain.LOGGER.log(LogLevel.ERROR, Text.translatable("meazy:file.output_cant_be_directory"));
            return null;
        }
        if (!outputFile.getParentFile().exists()) {
            if (outputFile.getParentFile().mkdirs()) {
                try {
                    if (outputFile.createNewFile()) MeazyMain.LOGGER.log(LogLevel.INFO, Text.translatable("meazy:file.created", args[1]));
                    else MeazyMain.LOGGER.log(LogLevel.INFO, Text.translatable("meazy:file.already_exists", args[1]));
                }
                catch (Exception e) {
                    MeazyMain.LOGGER.log(LogLevel.ERROR, Text.translatable("meazy:file.cant_create", args[1]));
                    return null;
                }
            }
            else {
                MeazyMain.LOGGER.log(LogLevel.ERROR, Text.translatable("meazy:file.cant_create_parent", args[1]));
                return null;
            }
        }

        String json = Registries.getGson().toJson(program, Program.class);
        try (FileWriter fileWriter = new FileWriter(outputFile)) {
            fileWriter.write(json);
        }
        catch (IOException e) {
            throw new RuntimeException(e);
        }

        return Text.translatable("meazy:commands.compile.info", (double) (endMillis - startMillis) / 1000);
    }
}

File: src/main/java/me/itzisonn_/meazy/command/custom/CompileCommand.kt
package me.itzisonn_.meazy.command.custom

import me.itzisonn_.meazy.command.AbstractCommand
import me.itzisonn_.meazy.text.Text
import me.itzisonn_.meazy.text.translatable
import me.itzisonn_.meazy.registry.Registries
import me.itzisonn_.meazy.logger.LogLevel
import me.itzisonn_.meazy.logger.Logger
import java.io.File
import java.io.IOException
import java.nio.file.Files

class CompileCommand : AbstractCommand(
    "compile",
    listOf("target_file", "output_directory_path")
) {
    override fun execute(vararg args: String): Text? {
        val file = File(args[0])
        if (file.isDirectory() || !file.exists()) {
            Logger.log(LogLevel.ERROR, translatable("meazy:file.doesnt_exist", file.absolutePath))
            return null
        }

        val extension = file.extension
        if (extension != "mea") {
            Logger.log(LogLevel.ERROR, translatable("meazy:file.unsupported_extension", extension))
            return null
        }

        Logger.log(
            LogLevel.INFO,
            translatable("meazy:commands.compile.compiling", file.absolutePath)
        )
        val startMillis = System.currentTimeMillis()

        val tokens = Registries.tokenizationFunction(file.readText())
        val program = Registries.parseTokensFunction(file, tokens)
        val classes = Registries.compileProgramFunction(program)

        val outputDirectory = File(args[1])
        if (!outputDirectory.exists()) {
            if (!outputDirectory.mkdirs()) {
                throw RuntimeException("Failed to create output directory") //TODO translation
            }
        }

        for ((classDesc, classFile) in classes) {
            val outputFile = File(outputDirectory, classDesc.displayName() + ".class")

            try {
                Files.write(outputFile.toPath(), classFile)
            }
            catch (e: IOException) {
                throw RuntimeException(e)
            }
        }

        val endMillis = System.currentTimeMillis()
        return translatable("meazy:commands.compile.info", (endMillis - startMillis).toDouble() / 1000)
    }
}
